src/basic/user-util.c

```diff
@@ -465,6 +465,54 @@ int in_group(const char *name) {
         return in_gid(gid);
 }
 
+static
+bool gid_list_has(const gid_t *list1, int size1, gid_t val) {
+        for (int i = 0; i < size1; i++)
+                if (list1[i] == val)
+                        return true;
+        return false;
+}
+
+int merge_gid_lists(const gid_t *list1, int size1, const gid_t *list2, int size2, gid_t **ret) {
+        int nresult = 0;
+        assert(ret);
+
+        if (size2 > INT_MAX - size1)
+                return -ENOBUFS;
+
+        gid_t *buf = new(gid_t, size1 + size2);
+        if (!buf)
+                return -ENOMEM;
+
+        for (int i = 0; i < size1; i++)
+                if (!gid_list_has(buf, nresult, list1[i]))
+                                buf[nresult++] = list1[i];
+        for (int i = 0; i < size2; i++)
+                if (!gid_list_has(buf, nresult, list2[i]))
+                                buf[nresult++] = list2[i];
+        *ret = buf;
+        return nresult;
+}
+
+int getgroups_alloc(gid_t** gids) {
+        gid_t *ids = NULL;
+        int ngids;
+
+        ngids = getgroups(0, NULL);
+        assert(ngids >= 0);
+
+        if (ngids > 0) {
+                ids = new(gid_t, ngids);
+                if (!ids)
+                        return -ENOMEM;
+
+                ngids = getgroups(ngids, ids);
+        }
+
+        *gids = ids;
+        return ngids;
```

Comment:
hmm, I'd prefer if this function would be implemented thread-safe, i.e. loop around getgroups() a couple of times, in case another thread manipulates the group list while we are on it. See in_group() how to do that...